### PR TITLE
Support script score when doc value is disabled and fix misusing DISI

### PR DESCRIPTION
**Pull request by bugmakerrrrrr**
### Description
The function was initially introduced in #1573, but was later reverted due to #1647. The root cause of the issue can be found in [this comment](https://github.com/opensearch-project/k-NN/issues/1647#issuecomment-2084495122). This PR reintroduces the function and fixes misusing DISI.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by navneet1v** [comment]
@bugmakerrrrrr can you add the entry for this feature in changelog.

**Comment by navneet1v** [comment]
@bugmakerrrrrr couple of things:
1. Can you add PR comment on the piece code that you have added to fix the bug. This will help me review the PR faster.
2. Did we add the IT which replicate the bug scenario to see the bug is not happening again. If possible can you add a comment in the PR for that too.

**Comment by ryanbogan** [comment]
Code looks good to me if we can add that integration test.  For the IT, the exact scenario was two vectors with data ingested, then run script score on either.

**Comment by navneet1v** [comment]
> Code looks good to me if we can add that integration test. For the IT, the exact scenario was two vectors with data ingested, then run script score on either.

can you paste the query and the index mapping @ryanbogan for better understanding

File: src/main/java/org/opensearch/knn/index/KNNVectorScriptDocValues.java [hunk 1]
```diff
@@ -6,30 +6,40 @@
...
-    private final BinaryDocValues binaryDocValues;
+    private final DocIdSetIterator vectorValues;
     private final String fieldName;
     @Getter
     private final VectorDataType vectorDataType;
     private boolean docExists = false;
+    private int lastDocID = -1;
 
     @Override
     public void setNextDocId(int docId) throws IOException {
```
Comment (bugmakerrrrrr) [inline src/main/java/org/opensearch/knn/index/KNNVectorScriptDocValues.java:31]: this method fix the bug in #1647

File: src/test/java/org/opensearch/knn/plugin/script/KNNScriptScoringIT.java [hunk 2]
```diff
@@ -669,7 +689,8 @@ private void testKNNScriptScore(SpaceType spaceType) throws Exception {
         final float[] queryVector = randomVector(dims);
         final BiFunction<float[], float[], Float> scoreFunction = getScoreFunction(spaceType, queryVector);
         for (String mapper : createMappers(dims)) {
-            createIndexAndAssertScriptScore(mapper, spaceType, scoreFunction, dims, queryVector);
+            createIndexAndAssertScriptScore(mapper, spaceType, scoreFunction, dims, queryVector, true);
+            createIndexAndAssertScriptScore(mapper, spaceType, scoreFunction, dims, queryVector, false);
```
Comment (bugmakerrrrrr) [inline src/test/java/org/opensearch/knn/plugin/script/KNNScriptScoringIT.java:693]: the issue in #1647 can be replicated using the non-dense test case.

**Review by ryanbogan — DISMISSED** [review]
LGTM! Thanks for the contribution!

**Review by jmazanec15 — DISMISSED** [review]
LGTM

**Comment by jmazanec15** [comment]
@bugmakerrrrrr can you fix conflict? Also, I have some good benchmarks for script scoring between lucene and plugin dv, Ill share those over on the issue.

**Comment by bugmakerrrrrr** [comment]
outcome

> @bugmakerrrrrr can you fix conflict? Also, I have some good benchmarks for script scoring between lucene and plugin dv, Ill share those over on the issue.

Fixed. Expecting to see these results.

**Comment by jmazanec15** [comment]
@bugmakerrrrrr Posted here: https://github.com/opensearch-project/k-NN/issues/1709.

**Review by jmazanec15 — APPROVED** [review]

**Review by ryanbogan — APPROVED** [review]

**Pull request merged**